jane/doc/extensions/uniqueness/intro.md

```diff
@@ -0,0 +1,121 @@
+In this document, we use the new [syntax for modes](../modes/syntax.md),
+which we expect to be stable soonish.
+
+The `unique` mode designates values that have only a single reference pointing
+to them. For example, an externally allocated data structure might support a
+`free` operation:
+
+```ocaml
+val free : t @ unique -> unit
+```
+
+Normally, a `free` operation would be dangerous since any remaining references
+can lead to a use-after-free segfault. But with the `unique` mode, the compiler
+guarantees that no further references remain. To provide this guarantee, the
+compiler tracks the uniqueness of values and marks values which have more than
+one reference pointing to them with the `aliased` mode:
+
+```ocaml
+let duplicate : t -> t * t @ aliased = fun t -> t, t
+```
+
+When a `unique` value is captured in a closure, this closure can be invoked at
+most once: if the closure was invoked more often, it could not use the value
+uniquely each time. Such closures are at mode `once`, while closures that can be
+invoked arbitrarily often are at mode `many`:
+
+```ocaml
+let delay_free : t @ unique -> (unit -> unit) @ once = fun t -> fun () -> free t
+```
+
+## Modalities
+
+These modes form two mode axis: the _uniqueness_ of a value is either `unique`
+or `aliased`, while the _affinity_ of a value is `once` or `many`. Similar to
+[locality](../local/intro.md), uniqueness and affinity are deep properties. If a
+value is at mode `unique` then all of its children are also `unique`. If a value
+is `once` then all of the closures it contains are also at mode `once`.
+
+We make an exception to this rule for record fields that are annotated with a
+_modality_. Analogous to global fields, a record field that is annotated as `@@
+aliased`, can contain `aliased` values even if the record itself is `unique`.
+However, a read from such a field always returns an `aliased` value. If a record
+field is annotated by `@@ many`, then it can only store `many` values even if
+the record itself is `once`. In return, a read of that field always yields a
+`many` value.
+
+## Mode Crossing
+
+You can always cast a value to a mode that affords fewer guarantees: You can use
+a `unique` value like an `aliased` value, and a `many` value like a `once`
+value. However, the other directions are disallowed in general: if a value is
+`aliased` there is no way to get it back to mode `unique`; similarly a `once`
+value can not be made `many`.
+
+We make an exception to this rule for values of those types that cross these
+modes. For example, `int`s and other immediates can be used as `unique` and
+`many`, no matter their mode. Similarly, most values that don't contain
+functions can be used as `many`. For example, an `int list @ once aliased` can
+be used as `many` but not as `unique`.
+
+## Checking for Uniqueness
+
+The compiler performs a sophisticated analysis to determine which values are
+aliased and which ones are unique. For example, it is fine to match on a value
```

Comment:
fine to match on a unique value and then use its contents uniquely

Reply:
No, it is `t` itself which is used here not its contents. The point I want to make here is that matching is not a use.

Reply:
ah right ok - I think this should still say 'match on a unique value and then use it' though